### PR TITLE
clone issues with netlify function...

### PR DESCRIPTION
**Pull request by berkeli**
## What does this change?

Module: n/a
Week(s): n/a

## Checklist

- [x] I have read the [contributing guidelines](CONTRIBUTING.MD)
- [x] I have checked my spelling and grammar with an [automated tool](https://www.grammarly.com/grammar-check)
- [x] I have previewed my changes to check the [markdown renders](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax) as I intend
- [x] I have run my code to check it works
- [x] My changes follow our [Style Guide](https://curriculum.codeyourfuture.io/guides/code-style-guide)

## Description

Adds clone issues functionality. Same as #182 but uses netlify functions for backend

The "clone all" button is commented out at the moment, because default netlify timeout is 10s and the function runs longer than that (due to numerous calls to github api, which are requried). The limit can be increased to 26s (max) via support ticket, which should be enough

Request to increase timeout: https://answers.netlify.com/t/function-timeout-increase/53881/483?u=berkeli

## Who needs to know about this?

@SallyMcGrath 

## Rendered Pages

See video in #182

### CODE REVIEW

File: netlify/helpers/github.ts [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+      })
+      .catch((error) => {
+        throw new Error(
+          `Could not get projects for ${this.login}, error: ${error.message}`
+        );
+      });
+
+    if (!project?.id) {
+      throw new Error(
+        "No project found that matches 'coursework planner' query, please clone the project template provided."
```
Comment (illicitonion) [inline netlify/helpers/github.ts:290]: I'm not quite sure as a user what this is asking me to do... Can you clarify?

It looks like you're looking for me to go to https://curriculum.codeyourfuture.io/fundamentals/prep/ and follow the "Set up Planner" instructions? Shall we link there?

**Review by illicitonion — COMMENTED** [review]
This generally looks great, thanks!

One overall theme is that errors seem to get logged straight to the netlify console rather than presented to users - can we make it so that errors show up for users? Fine if it's a horrible "Un-styled black text on white background" error, but giving the message to the user seems important?

I have a few minor clean-ups I'll PR to you, but nothing very interesting.

As I DM'd, the projects GraphQL query seems to be returning more `null`-like data than I would've expected - let's debug offline... But I'd like to see this working myself before merging :)

Oh and one more question - what action takes >10 seconds? Is it a "Clone all issues for the whole course" action or something? I would expect any one module to be able to be cloned in <10 seconds?

**Comment by illicitonion** [comment]
I filed https://github.com/CodeYourFuture/curriculum/pull/262 against this branch with a few minor clean-ups :)

File: netlify/helpers/const.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import config from "./config";
+
+export const GITHUB_DEFAULT_LABELS = [
```
[the remark this answers is not in the document]
Comment (berkeli) [reply netlify/helpers/const.ts:3]: It's another call to the api, which makes it even slower :)

File: netlify/helpers/util.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+export function getTokenFromCookies(headers: HandlerEvent["headers"]): string {
+  if (
+    headers === null ||
+    headers === undefined ||
+    headers.cookie === undefined
+  ) {
+    return "";
+  }
+
+  // Split a cookie string in an array (Originally found http://stackoverflow.com/a/3409200/1427439)
```
[the remark this answers is not in the document]
Comment (berkeli) [reply netlify/helpers/util.ts:107]: The typings seems to be different from one provider to another for `HandlerEvent`, which is probably why :)

File: netlify/functions/clone.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+  // check for state
+  // this endpoint is only to be used for cloning issues, so we should ignore everything that doesn't have state (see type for state).
+  const stateOrFail = checkState(event);
+  if ("statusCode" in stateOrFail) {
+    return withHeaders(stateOrFail);
+  }
+
+  // check for auth token in cookies
+  token = getTokenFromCookies(event.headers);
```
[the remark this answers is not in the document]
Comment (berkeli) [reply netlify/functions/clone.ts:30]: No need to pre-declare, thanks

File: layouts/partials/page-header.html [hunk 5]
```diff
@@ -34,13 +34,20 @@ <h1 class="c-page-header__title e-heading__1">
...
+      <h2
+        class="c-page-header__subtitle c-page-header__backlog e-heading__3"
+        style="display:flex; width:max-content;">
         <a
           class="c-page-header__link"
           href="{{ .Page.Site.Params.org }}/{{ .Page.Params.backlog }}/issues"
           >{{ .Page.Params.backlog }}
           🔗</a
         >
+        <!-- <a
```
[the remark this answers is not in the document]
Comment (berkeli) [reply layouts/partials/page-header.html:46]: Yeah we can remove it, although it should work once we request netlify to increase timeout from 10s to 26s

File: netlify/helpers/github.ts (hunk 1 above)
Comment (berkeli) [reply netlify/helpers/github.ts:290]: Thanks, I will add the link

**Comment by berkeli** [comment]
> Oh and one more question - what action takes >10 seconds? Is it a "Clone all issues for the whole course" action or something? I would expect any one module to be able to be cloned in <10 seconds?

It's the clone all button, which I had left commented out. Basically getting repo + project information, then cloning issues. I've tried making "parallel" api calls where possible but still quite slow.

I'll file a request with netlify to increase the timeout which should allow us to clone all

**Comment by berkeli** [comment]
I have also requested timeout increase for netlify function

https://answers.netlify.com/t/function-timeout-increase/53881/483?u=berkeli

**Review by illicitonion — APPROVED** [review]
Amazing, thanks!

**Pull request merged**